rust/cardano-blockchain-types/src/cip36/mod.rs

```diff
@@ -0,0 +1,143 @@
+//! CIP-36 Catalyst registration module
+
+pub mod key_registration;
+pub mod registration_witness;
+mod validation;
+pub mod voting_pk;
+
+use ed25519_dalek::VerifyingKey;
+use key_registration::Cip36KeyRegistration;
+use pallas::ledger::addresses::ShelleyAddress;
+use registration_witness::Cip36RegistrationWitness;
+use validation::{validate_payment_address_network, validate_signature, validate_voting_keys};
+use voting_pk::VotingPubKey;
+
+use crate::{MetadatumValue, Network};
+
+/// CIP-36 Catalyst registration
+#[derive(Clone, Default, Debug)]
+pub struct Cip36 {
+    /// Key registration - 61284
+    key_registration: Cip36KeyRegistration,
+    /// Registration witness - 61285
+    registration_witness: Cip36RegistrationWitness,
+    /// Is this a Catalyst strict registration?
+    is_catalyst_strict: bool,
+}
+
+/// Validation value for CIP-36.
+#[allow(clippy::struct_excessive_bools, clippy::module_name_repetitions)]
+#[derive(Clone, Default, Debug)]
+pub struct Cip36Validation {
+    /// Is the signature valid? (signature in 61285)
+    pub is_valid_signature: bool,
+    /// Is the payment address on the correct network?
+    pub is_valid_payment_address_network: bool,
+    /// Is the voting keys valid?
+    pub is_valid_voting_keys: bool,
+    /// Is the purpose valid? (Always 0 for Catalyst)
+    pub is_valid_purpose: bool,
+}
+
+impl Cip36 {
+    /// Create an instance of CIP-36.
+    #[must_use]
+    pub fn new(
+        key_registration: Cip36KeyRegistration, registration_witness: Cip36RegistrationWitness,
```

Comment:
We should not require the caller to supply anything more than the `MultiEraBlock` and a `TxnIndex` to be decoded.
The `new` should be fallible and fail if it can not detect any CIP36 metadata in the transaction of the block at that index, or succeed if there is, but return a Cip36 structure that contains a non empty error report if there looked like there was CIP36 metadata, but it was not able to be validated.

The `is_catalyst_strict` option is still valid, so the paramters should be `multi_era_block, txn_index, strict`

Reply:
The caller should not be calling the decode directly on the individual metadatums themselves, it should all be encapsulated here.